### PR TITLE
Correct repeat panel

### PR DESCRIPTION
**Pull request by ilgizar**
Issue #10553

### CODE REVIEW

File: public/app/features/dashboard/dashboard_model.ts [hunk 1]
```diff
@@ -370,6 +387,40 @@ export class DashboardModel {
...
+    // Update gridPos for panels below
+    if (selectedOptions.length > 1) {
+      let panelBelowIndex = panelIndex + selectedOptions.length;
+      for (let i = panelBelowIndex; i < this.panels.length; i++) {
+        this.panels[i].gridPos.y += yPos;
+      }
+    }
+  }
+
+  repeatCollapsedPanel(panel: PanelModel, panelIndex: number, repeatPanel: PanelModel) {
```
Comment (torkelo) [inline public/app/features/dashboard/dashboard_model.ts]: why do we need to repeat collapsed panels?

File: public/app/features/dashboard/dashboard_model.ts [hunk 2]
```diff
@@ -272,24 +272,41 @@ export class DashboardModel {
       let panel = this.panels[i];
       if (panel.repeat) {
         this.repeatPanel(panel, i);
+      } else if (panel.panels) {
+        // find repeate panel on collapsed state
+        let repeatPanel;
+        for (let j = 0; j < panel.panels.length; j++) {
+          if (panel.panels[j].repeat) {
+            repeatPanel = panel.panels[j];
```
Comment (torkelo) [inline public/app/features/dashboard/dashboard_model.ts]: this also looks strange, why just a repeat a single panel in the collapsed row?

Comment (ilgizar) [reply public/app/features/dashboard/dashboard_model.ts]: I agree. There can be several such panels in a collapsed row. This is my omission - I'll fix it.

File: public/app/features/dashboard/dashboard_model.ts (hunk 1 above)
Comment (ilgizar) [reply public/app/features/dashboard/dashboard_model.ts]: To recalc hidden panels and correct view number on collanped row. To show on First row "(1 hidden panels)" instead "(2 hidden panels)" in this situation

**Comment by ilgizar** [comment]
On view template file https://github.com/grafana/grafana/blob/4720b86f5c1352360402615260e0649eb510f441/public/app/features/dashboard/dashgrid/DashboardRow.tsx#L82
and show it
https://github.com/grafana/grafana/blob/4720b86f5c1352360402615260e0649eb510f441/public/app/features/dashboard/dashgrid/DashboardRow.tsx#L89

As an option, I can store number of panels to variable. And display it instead of real number of panels.

**Comment by torkelo** [comment]
I do not think that hidden number count needs to reflect the current repeat panel count.

**Comment by ilgizar** [comment]
This is a question of usability.
First situation: The variable is set to "all" and is updated when the dashboard is loaded. How many panels hide on collapsed row: 10, 100, 1000? if you open a collapsed row with thousands of panels, then the browser tab can be safely killed until it has eaten all the memory.
Second situation: I have collapsed row with thousands of panels, because the variable is set to "all". I filter out the elements according to the necessary condition, and I want to see the number in the header of the row, and not just in the variable menu.
Does I convinced you that you need to display the correct number of panels in a collapsed line (without cloning, only counting)?

**Comment by torkelo** [comment]
no, I do not think you should even have an "all" option for a variable with 1000 options that is used for repeating.

**Comment by torkelo** [comment]
As I said, all options does not look appropriate for a variable that have this many options (especially one that is used in a repeating scenario )

**Comment by ilgizar** [comment]
Ok, I will delete my code correcting the number of panels on the collapsed row.

File: public/app/features/dashboard/dashboard_model.ts [hunk 3]
```diff
@@ -588,6 +600,10 @@ export class DashboardModel {
         }
 
         row.panels = [];
+
+        if (needRepeats) {
+          this.processRepeats();
```
Comment (torkelo) [inline public/app/features/dashboard/dashboard_model.ts]: this seems recursive, can you explain why this needed?

**Review by torkelo — CHANGES_REQUESTED** [review]
A unit test for the scenario your trying to solve would breat, you find all the repeat unit tests here: https://github.com/grafana/grafana/blob/master/public/app/features/dashboard/specs/repeat.jest.ts

**Comment by torkelo** [comment]
fixed by https://github.com/grafana/grafana/pull/10619

**Pull request closed without merging**